waku/node/peer_manager/peer_manager.nim

```diff
@@ -115,21 +119,38 @@ proc addPeer*(pm: PeerManager, remotePeerInfo: RemotePeerInfo, origin = UnknownO
     # Do not attempt to manage our unmanageable self
     return
 
-  # ...public key
-  var publicKey: PublicKey
-  discard remotePeerInfo.peerId.extractPublicKey(publicKey)
-
   if pm.peerStore[AddressBook][remotePeerInfo.peerId] == remotePeerInfo.addrs and
-     pm.peerStore[KeyBook][remotePeerInfo.peerId] == publicKey and
+     pm.peerStore[KeyBook][remotePeerInfo.peerId] == remotePeerInfo.publicKey and
      pm.peerStore[ENRBook][remotePeerInfo.peerId].raw.len > 0:
     # Peer already managed and ENR info is already saved
     return
 
   trace "Adding peer to manager", peerId = remotePeerInfo.peerId, addresses = remotePeerInfo.addrs
-
+    
   pm.peerStore[AddressBook][remotePeerInfo.peerId] = remotePeerInfo.addrs
-  pm.peerStore[KeyBook][remotePeerInfo.peerId] = publicKey
+  pm.peerStore[KeyBook][remotePeerInfo.peerId] = remotePeerInfo.publicKey
   pm.peerStore[SourceBook][remotePeerInfo.peerId] = origin
+  
+  if remotePeerInfo.protocols.len > 0:
+    pm.peerStore[ProtoBook][remotePeerInfo.peerId] = remotePeerInfo.protocols
+  elif remotePeerInfo.enr.isSome():
+    # RemotePeerInfo from discv5 always have empty protocol list
+    let caps = remotePeerInfo.enr.get().getCapabilities()
+    
+    var protos: seq[string]
+    for cap in caps:
+      case cap:
+        of Relay:
+          protos.add(WakuRelayCodec)
+        of Store:
+          protos.add(WakuStoreCodec)
+        of Filter:
+          protos.add(WakuFilterSubscribeCodec)
+          protos.add(WakuFilterPushCodec)
+        of LightPush:
+          protos.add(WakuLightPushCodec)
+
+    pm.peerStore[ProtoBook][remotePeerInfo.peerId] = protos
```

Comment:
We have a chicken and egg problem here. If the protos are not populated we can't filter non-Relay peers then we don't connect and don't get the protos from identify.

Alternative would be to connect to all peer we discover regardless of their (ENR) capability.

Reply:
Indeed. We shouldn't mix the two concepts here:
- the Protobook tracks what libp2p protocols you have _identified_ peers as supporting
- ENR capabilities provide an optional, "convenience flag" to filter for interesting peers before having to connect

Nodes may support more protocols than they're willing to advertise in their ENRs (which may be their way to say "I support Store, but not for you") and they may even lie about what they support. TL;DR the ENR protocols should be used for initial filtering, but should not be seen as authoritative on the libp2p level.

Reply:
(That said, I can see the complexity this would cause elsewhere as we do not pre-filter on protocol when adding the peers, only when maintaining the connections. It may be possible to retrieve the ENR from the ENR book at the point of retrieving relay peers to connect to, but I understand if there are other complexities that make this a necessary workaround.)

Reply:
As soon as you connect the protos gets overwritten by identify anyway.

You either have to remember that protos are never populated before the first connection and the ENR caps should be used OR remember that before the first connection the protos maybe not be the same as after.

I feel like the second case is less of a gotcha than the first.